When a layer size changes, recompute compositing from its stacking context down

squelart opened the pull request:
#### 7b1e2f8cef2e6be5d52dd160ff14025bf8e592ae
<pre>
When a layer size changes, recompute compositing from its stacking context down
<a href="https://bugs.webkit.org/show_bug.cgi?id=276216">https://bugs.webkit.org/show_bug.cgi?id=276216</a>
<a href="https://rdar.apple.com/130615925">rdar://130615925</a>

Reviewed by Matt Woodrow.

In some cases, changing the size of an element could change how layers overlap,
so the compositing around it should be appropriately recomputed.

* LayoutTests/compositing/shared-backing/update-backing-sharing-on-geometry-change-expected.html: Added.
* LayoutTests/compositing/shared-backing/update-backing-sharing-on-geometry-change.html: Added.
Before this patch, the changing geometry of the container, which moves the red
square into/out of the yellow box, would not correctly update the green box.

* LayoutTests/compositing/shared-backing/update-backing-sharing-on-size-change-expected.html: Added.
* LayoutTests/compositing/shared-backing/update-backing-sharing-on-size-change.html: Added.
Before this patch, this reduced test case was triggering `ASSERT(provider)` in
RenderLayerCompositor::traverseUnchangedSubtree.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateLayerPosition):

Canonical link: <a href="https://commits.webkit.org/280673@main">https://commits.webkit.org/280673@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7d4f5c735a008bec81a453785c91497c9623688e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57287 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36615 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9762 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60909 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7730 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59415 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44239 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7920 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46384 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5452 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59317 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34359 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49468 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27247 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31141 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6781 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6735 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/53101 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7052 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62588 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1200 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7144 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53644 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1205 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49504 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53721 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12662 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1016 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32444 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33529 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34614 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33275 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on LayoutTests/compositing/shared-backing/bug276216.html:
```diff
@@ -0,0 +1,18 @@
+<!DOCTYPE html> <!-- webkit-test-runner [ OverlappingBackingStoreProvidersEnabled=true ] -->
```
This hopefully isn't really necessary (preferences that are 'testable' or higher should be enabled by default).

mattwoodrow reviewed (approved):
Nice catch!

smfr commented on LayoutTests/compositing/shared-backing/bug276216.html:
```diff
@@ -0,0 +1,18 @@
+<!DOCTYPE html>
```
I don't like test cases with bug numbers in the name because it's hard to know what they do. Let's call it `update-backing-sharing-on-size-change.html` or something.

smfr replied:
If this fixes https://bugs.webkit.org/show_bug.cgi?id=275937 please also make a test case from its attachment and include it here.

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -1732,7 +1732,9 @@ bool RenderLayer::updateLayerPosition(OptionSet<UpdateLayerPositionsFlag>* flags
     auto layerRect = computeLayerPositionAndIntegralSize(renderer());
     auto localPoint = layerRect.location();
 
+    bool footprintChanged = false;
```
Not a fan of "footprint" (we normally use it in the context of memory). Maybe `geometryChanged` or `sizeOrPositionChanged`.

squelart commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -1793,29 +1795,30 @@ bool RenderLayer::updateLayerPosition(OptionSet<UpdateLayerPositionsFlag>* flags
         if (auto* boxModelObject = dynamicDowncast<RenderBoxModelObject>(renderer())) {
             auto newOffset = boxModelObject->offsetForInFlowPosition();
-            positionOrOffsetChanged = newOffset != m_offsetForPosition;
+            geometryChanged = newOffset != m_offsetForPosition;
```
Whoops, this should be `|=`, so we don't stomp on a previous `= true`.

squelart committed:
When a layer size changes, recompute compositing from its stacking context down
https://bugs.webkit.org/show_bug.cgi?id=276216
rdar://130615925

Reviewed by Matt Woodrow.

In some cases, changing the size of an element could change how layers overlap,
so the compositing around it should be appropriately recomputed.

* LayoutTests/compositing/shared-backing/update-backing-sharing-on-geometry-change-expected.html: Added.
* LayoutTests/compositing/shared-backing/update-backing-sharing-on-geometry-change.html: Added.
Before this patch, the changing geometry of the container, which moves the red
square into/out of the yellow box, would not correctly update the green box.

* LayoutTests/compositing/shared-backing/update-backing-sharing-on-size-change-expected.html: Added.
* LayoutTests/compositing/shared-backing/update-backing-sharing-on-size-change.html: Added.
Before this patch, this reduced test case was triggering `ASSERT(provider)` in
RenderLayerCompositor::traverseUnchangedSubtree.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateLayerPosition):

Canonical link: https://commits.webkit.org/280673@main
